numpy/core/src/multiarray/descriptor.c

```diff
@@ -3380,12 +3384,19 @@ static PyMappingMethods descr_as_mapping = {
 };
 
 /****************** End of Mapping Protocol ******************************/
+#ifdef USE_DTYPE_AS_PYOBJECT
+#pragma message "USE_DTYPE_AS_PYOBJECT defined, using old dtypes"
+#define BASETYPE 0
+#else
+#pragma message "USE_DTYPE_AS_PYOBJECT not defined, using new dtypes"
+#define BASETYPE &PyType_Type
+#endif
 
 NPY_NO_EXPORT PyTypeObject PyArrayDescr_Type = {
 #if defined(NPY_PY3K)
-    PyVarObject_HEAD_INIT(NULL, 0)
+    PyVarObject_HEAD_INIT(BASETYPE, 0)
```

Comment:
I think there's a bug in various compilers that means it's not actually legal to take the address of `PyType_Type` here - in CPython, I've seen cases where they have to manually set it later.

Reply:
I also suspect this is not necessary, the default is always for this field to be `PyType_Type`

Reply:
> bug in various compilers that means it's not actually legal to take the address

Seems [you are right](https://stackoverflow.com/questions/45401933) and MSVC (surprise, surprise) is the offending compiler

Reply:
> this is not necessary

adopted